presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarFromAnnotationsParser.java

```diff
@@ -99,11 +106,37 @@ private static List<ScalarHeaderAndMethods> findScalarsInFunctionSetClass(Class<
         return builder.build();
     }
 
+    private static Optional<ScalarStatsHeader> getScalarStatsHeader(Method annotated)
+    {
+        Optional<ScalarStatsHeader> scalarStatsHeader;
+        ScalarFunctionConstantStats constantStatsAnnotation =
+                annotated.getAnnotation(ScalarFunctionConstantStats.class);
+        List<java.lang.reflect.Parameter> params =
+                Arrays.stream(annotated.getParameters())
+                        .filter(param -> param.getAnnotation(SqlType.class) != null)
+                        .collect(Collectors.toList());
+        // Map of (function argument position index) -> (ScalarPropagateSourceStats annotation)
+        Map<Integer, ScalarPropagateSourceStats> argumentStatsResolver = new HashMap<>();
+
+        IntStream.range(0, params.size())
+                .filter(paramIndex -> params.get(paramIndex).getAnnotation(ScalarPropagateSourceStats.class) != null)
+                .forEachOrdered(paramIndex -> argumentStatsResolver.put(paramIndex,
+                        params.get(paramIndex).getAnnotation(ScalarPropagateSourceStats.class)));
+
+        scalarStatsHeader = Optional.ofNullable(constantStatsAnnotation)
+                .map(statsAnnotation -> new ScalarStatsHeader(statsAnnotation, argumentStatsResolver));
+        if (!argumentStatsResolver.isEmpty() && !scalarStatsHeader.isPresent()) {
+            scalarStatsHeader = Optional.of(new ScalarStatsHeader(argumentStatsResolver));
+        }
+        return scalarStatsHeader;
+    }
+
     private static SqlScalarFunction parseParametricScalar(ScalarHeaderAndMethods scalar, Optional<Constructor<?>> constructor)
     {
         ScalarImplementationHeader header = scalar.getHeader();
 
         Map<SpecializedSignature, ParametricScalarImplementation.Builder> signatures = new HashMap<>();
+        Map<Signature, ScalarStatsHeader> signatureVsStatsHeader = new HashMap<>();
```

Comment:
We prefer the use of immutable collections, can we switch this to use a `ImmutableMap.builder()`

Reply:
Let's rename this to `signatureToStatsHeader`